orm/iterator.go

```diff
@@ -92,6 +95,91 @@ func First(it Iterator, dest Persistent) (RowID, error) {
 	return binKey, nil
 }
 
+func Paginate(
+	it Iterator,
+	pageRequest *query.PageRequest,
+	dest ModelSlicePtr,
+) (*query.PageResponse, error) {
+	// if the PageRequest is nil, use default PageRequest
+	if pageRequest == nil {
+		pageRequest = &query.PageRequest{}
+	}
+
+	offset := pageRequest.Offset
+	key := pageRequest.Key
+	limit := pageRequest.Limit
+	countTotal := pageRequest.CountTotal
+
+	if offset > 0 && key != nil {
+		return nil, fmt.Errorf("invalid request, either offset or key is expected, got both")
+	}
+
+	if limit == 0 {
+		limit = 100
+
+		// count total results when the limit is zero/not supplied
+		countTotal = true
+	}
+
+	if it == nil {
+		return nil, errors.Wrap(ErrArgument, "iterator must not be nil")
+	}
+	defer it.Close()
+
+	var slice, t reflect.Value
+	typ, err := assertDest(dest, &slice, &t)
+	if err != nil {
+		return nil, err
+	}
+
+	end := offset + limit
+
+	var count uint64
+	var nextKey []byte
+	for {
+		obj := reflect.New(typ)
+		val := obj.Elem()
+		model := obj
```

Comment:
if `elemType` is already a pointer (e.g. `dest` being some pointer to a slice of pointers, like `[]*GroupMember`), then `obj` is a pointer to a pointer which will cause issues when doing `obj.Interface().(codec.ProtoMarshaler)`. For that reason, we copy `obj` into `model` if we have a simple pointer but in case `elemType.Kind() == reflect.Ptr`, we overwrite it with `model = val`.